infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -175,231 +186,272 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
-        log.trace("updateCommittedTail: received request to update committed tail {}", msg);
-        streamLog.updateCommittedTail(msg.getPayload());
-        r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+    @RequestHandler(type = PayloadCase.UPDATE_COMMITTED_TAIL_REQUEST)
+    public void handleUpdateCommittedTailRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleUpdateCommittedTailRequest: received request to "
+                    + "update committed tail {}", TextFormat.shortDebugString(req));
+        }
+
+        streamLog.updateCommittedTail(req.getPayload().getUpdateCommittedTailRequest().getCommittedTail());
+        HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+        ResponseMsg response = getResponseMsg(responseHeader, getUpdateCommittedTailResponseMsg());
+        r.sendResponse(response, ctx);
     }
 
     /**
      * A helper function that maps an exception to the appropriate response message.
      */
-    private void handleException(Throwable ex, ChannelHandlerContext ctx, CorfuPayloadMsg msg, IServerRouter r) {
-        log.trace("handleException: handling exception {} for {}", ex, msg);
+    private void handleException(Throwable ex, ChannelHandlerContext ctx, RequestMsg req, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleException: handling exception {} for {}", ex, TextFormat.shortDebugString(req));
+        }
+
+        HeaderMsg responseHeader;
+
         if (ex.getCause() instanceof WrongEpochException) {
             WrongEpochException wee = (WrongEpochException) ex.getCause();
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, wee.getCorrectEpoch()));
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getWrongEpochErrorMsg(wee.getCorrectEpoch())), ctx);
         } else if (ex.getCause() instanceof OverwriteException) {
             OverwriteException owe = (OverwriteException) ex.getCause();
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE
-                    .payloadMsg(owe.getOverWriteCause().getId()));
-        } else if (ex.getCause() instanceof DataOutrankedException) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } else if (ex.getCause() instanceof ValueAdoptedException) {
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getOverwriteErrorMsg(owe.getOverWriteCause().getId())), ctx);
+        } else if (ex.getCause() instanceof  DataOutrankedException) {
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getDataOutrankedErrorMsg()), ctx);
```

Comment:
Since the same request header is used for the response header these checks won't make a  lot of sense, right?

Imagine the client is on epoch 1 and sends a message to the server and the server transitions to epoch 2 just before it sends out the response. When the client receives the response it will be on epoch 1, but ignoreEpoch=false would mean that the client should have raised an exception. In this case it wont.

Reply:
It's a semantic change from the current code base. There is an difference between sending back the wrong epoch and saying that the return epoch value has no effect on the final result. That difference should be explicitly encoded.

Reply:
With PR [2855](https://github.com/CorfuDB/CorfuDB/pull/2855), this message type is no longer needed and has been removed.